### PR TITLE
Create default forms on installation or update

### PR DESCRIPTION
**Pull request by AdrienClairembault**
## Checklist before requesting a review

*Please delete options that are not relevant.*

- [X] I have performed a self-review of my code.
- [X] I have added tests that prove my fix is effective or that my feature works.

## Description

Add two default forms when installing GLPI or updating to GLPI 11.
The goal is to copy the "old" helpdesk form:

image

It is split into two Incident/Request forms which are almost identical.
Note that the "associated elements" question is missing because it isn't possible to implement it right now (we will work on that later) and that the picked "watchers" are also not applied into the ticket (not possible yet too).

The default forms themselves are subject to change later as we make the final adjustments before GLPI 11.
Right now, the goal is just to offer the same feature that in GLPI 10 but with two "real" forms.

### CODE REVIEW

**Comment by AdrienClairembault** [comment]
Update tests fail with MariaDB, probably because it doesn't compare json correctly (as JSON is only a real type in MySQL).
We probably need to use the dedicated json_equals somewhere if we are on MariaDB but I am not sure if we want to waste time working on that (for a false positive).

image

Suggestions are welcome.

**Comment by AdrienClairembault** [comment]
Nvm it is actually MySQL that fails the comparison 🤷 
I'm tempted to just ignore these two json fields in the update tests, ok for you @cedric-anne ?

**Comment by trasher** [comment]
> I'm tempted to just ignore these two json fields in the update tests

I guess the problem is the tests suite is not ready for that kind of data; and therefore should be fixed; not ignored. Anyway, I do not know well that part, I've no proposition so far.

**Comment by AdrienClairembault** [comment]
> I guess the problem is the tests suite is not ready for that kind of data; and therefore should be fixed; not ignored. Anyway, I do not know well that part, I've no proposition so far.

I agree but we have to pick our priorities at this time and this specific false positive case is very low in my TODO list.
If anyone else has more free time, feel free to correct it.

File: src/Glpi/Form/DefaultFormsManager.php [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+            $this->getDescriptionQuestionData(),
+        ]);
+
+        // Find title and description question tags, needed to configure the created ticket
+        $title_tag = $this->answer_tag_provider->getTagForQuestion(
+            $form->getNthQuestion(4)
+        );
+        $description_tag = $this->answer_tag_provider->getTagForQuestion(
+            $form->getNthQuestion(5)
+        );
```
Comment (cedric-anne) [inline src/Glpi/Form/DefaultFormsManager.php:105]: This piece of code is not really resilient to default form data changes. It could probably be improved.

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Review by cedric-anne — APPROVED** [review]

**Review by orthagh — APPROVED** [review]
A few remark after I tested the current pr:

- the first one is already noted:
   > Note that the "associated elements" question is missing

   We should not ship 11.0 without that.

- Reset forms on an existing instance (update case). I found it worked when I removed all my existing forms. Maybe we can do better here? A cli command, a button in debug mode ?

- The current render of the form (by previewing it): it's bad; especially in comparison of the current hard-coded form. On my QHD screen, it's not displayed entirely without scrolling. Do we need to have one container with bug margins/paddings for each question/inputs?

**Comment by AdrienClairembault** [comment]
> The current render of the form (by previewing it): it's bad;

This is known and will be worked on soon, don't worry ;)

**Pull request merged**